File: internal/playlist/generator.go
package playlist

import (
	"context"
	"fmt"
	"strconv"
	"strings"
	"time"

	"sonos-playlist/internal/ai"
	"sonos-playlist/internal/output"
	"sonos-playlist/internal/sonos"
	"sonos-playlist/internal/storage"
)

type GeneratorOptions struct {
	Keys             ai.APIKeys
	Prompt           string
	Room             string
	DryRun           bool
	Client           *sonos.Client
	Monitor          bool
	CountPerProvider int
	Output           *output.Output
}

type Result struct {
	Room            string `json:"room"`
	DryRun          bool   `json:"dryRun"`
	QueuedSongs     int    `json:"queuedSongs"`
	FailedSongs     int    `json:"failedSongs"`
	PlaybackStarted bool   `json:"playbackStarted"`
	Monitored       bool   `json:"monitored"`
}

const (
	minSongs         = 30
	maxRetries       = 5
	altRetryCooldown = 30 * time.Second
	fastStartTimeout = 10 * time.Second
)

func songKey(song ai.Song) string {
	return strings.ToLower(song.Artist) + ":::" + strings.ToLower(song.Title)
}

func generateFastStartSongs(ctx context.Context, keys ai.APIKeys, prompt string) []ai.Song {
	if keys.Google == "" {
		return []ai.Song{}
	}
	fctx, cancel := context.WithTimeout(ctx, fastStartTimeout)
	defer cancel()
	songs, err := ai.GeneratePlaylistGemini(fctx, keys.Google, prompt, 3)
	if err != nil {
		return []ai.Song{}
	}
	return songs
}

func GenerateAndPlay(ctx context.Context, options GeneratorOptions) (Result, error) {
	keys := options.Keys
	prompt := options.Prompt
	room := options.Room
	dryRun := options.DryRun
	client := options.Client
	monitor := options.Monitor
	countPerProvider := options.CountPerProvider
	out := options.Output

	playbackStarted := false
	var fastStartSong *ai.Song
	existingKeys := map[string]struct{}{}
	triedTrackIDsBySong := map[string]map[int]struct{}{}
	lastAltAttemptAt := map[string]time.Time{}
	var monitorHandle *sonos.MonitorHandle

	handleUnplayable := func(song ai.Song, trackID string) error {
		key := songKey(song)
		if trackID != "" {
			out.Warn(fmt.Sprintf("Detected unplayable track: %s - %s (ID: %s, now blocked)", song.Artist, song.Title, trackID))
		} else {
			out.Warn(fmt.Sprintf("Detected unplayable track: %s - %s", song.Artist, song.Title))
		}
		now := time.Now()
		if last, ok := lastAltAttemptAt[key]; ok && now.Sub(last) < altRetryCooldown {
			return nil
		}
		lastAltAttemptAt[key] = now

		tried, ok := triedTrackIDsBySong[key]
		if !ok {
			tried = map[int]struct{}{}
			triedTrackIDsBySong[key] = tried
		}
		if trackID != "" {
			if id, err := strconv.Atoi(trackID); err == nil {
				tried[id] = struct{}{}
			}
		}

		alt := sonos.AddAlternateSongToQueue(ctx, client, room, song, tried)
		if alt.Success && alt.TrackID != 0 {
			tried[alt.TrackID] = struct{}{}
			if trackID != "" {
				storage.SetTrackReplacement(trackID, strconv.Itoa(alt.TrackID))
			}
			out.Warn(fmt.Sprintf("Unavailable track replaced with alternate: %s - %s", song.Artist, song.Title))
		} else {
			out.Warn(fmt.Sprintf("No alternate found for: %s - %s", song.Artist, song.Title))
		}
		return nil
	}

	ensureMonitor := func() {
		if !monitor || monitorHandle != nil {
			return
		}
		h := sonos.StartPlaybackMonitor(ctx, client, room, sonos.MonitorOptions{
			OnUnplayable: handleUnplayable,
			UseTimer:     false,
		})
		monitorHandle = &h
		out.Info(out.Gray("Monitoring playback for unavailable tracks (Ctrl+C to stop)..."))
	}

	var songs []ai.RankedSong
	if !dryRun {
		out.Info("Generating fast-start songs (Gemini 3 Flash)...")
		fastStartCandidates := generateFastStartSongs(ctx, keys, prompt)

		if err := sonos.Pause(ctx, client, room); err != nil {
			// ignore pause failures
		}
		if err := sonos.ClearQueue(ctx, client, room); err != nil {
			return Result{}, err
		}

		if len(fastStartCandidates) == 0 {
			out.Warn("Fast-start unavailable; continuing")
		} else {
			for _, candidate := range fastStartCandidates {
				result := sonos.AddSongToQueue(ctx, client, room, candidate, true)
				if !result.Success {
					continue
				}
				key := songKey(candidate)
				if result.TrackID != 0 {
					tried := triedTrackIDsBySong[key]
					if tried == nil {
						tried = map[int]struct{}{}
						triedTrackIDsBySong[key] = tried
					}
					tried[result.TrackID] = struct{}{}
				}
				existingKeys[key] = struct{}{}
				playbackStarted = true
				c := candidate
				fastStartSong = &c
				out.Success(fmt.Sprintf("  -> Now playing: %s - %s", candidate.Artist, candidate.Title))
				ensureMonitor()
				break
			}
			if !playbackStarted {
				out.Warn("Fast-start songs queued but none were playable")
			}
		}
	}

	out.Info(fmt.Sprintf("Generating playlist from all providers for: %q", prompt))
	ranked, err := ai.GeneratePlaylistAllProviders(ctx, keys, prompt, countPerProvider)
	if err != nil {
		return Result{}, err
	}
	songs = ranked
	out.Success(fmt.Sprintf("Generated %d unique songs from all providers", len(songs)))

	out.Print(out.Bold("Initial playlist:"))
	for i, song := range songs {
		voteIndicator := ""
		if song.Votes > 1 {
			voteIndicator = out.Yellow(fmt.Sprintf(" [%d]", song.Votes))
		}
		out.Print(fmt.Sprintf("  %d. %s - %s%s", i+1, song.Artist, song.Title, voteIndicator))
	}
	out.Print("")

	if dryRun {
		out.Warn("Dry run - not queueing or playing")
		return Result{Room: room, DryRun: true, QueuedSongs: 0, FailedSongs: 0, PlaybackStarted: false, Monitored: false}, nil
	}

	if playbackStarted {
		out.Info("Fast-start playing; appending generated songs to the queue...")
	}

	queuedSongs := []ai.Song{}
	failedSongs := []ai.Song{}
	retryCount := 0

	for len(queuedSongs) < minSongs && retryCount < maxRetries {
		if retryCount > 0 {
			out.Info(fmt.Sprintf("Generating more songs (attempt %d)...", retryCount+1))
			existing := make([]ai.Song, 0, len(queuedSongs)+len(failedSongs)+1)
			existing = append(existing, queuedSongs...)
			existing = append(existing, failedSongs...)
			if fastStartSong != nil {
				existing = append(existing, *fastStartSong)
			}
			moreSongs, err := ai.GenerateMoreSongs(ctx, keys, prompt, existing, countPerProvider)
			if err != nil {
				return Result{}, err
			}
			if len(moreSongs) == 0 {
				out.Warn("Could not generate more unique songs")
				break
			}
			songs = make([]ai.RankedSong, 0, len(moreSongs))
			for _, s := range moreSongs {
				songs = append(songs, ai.RankedSong{Song: s, Votes: 1})
			}
			out.Success(fmt.Sprintf("Generated %d more songs", len(songs)))
		}

		out.Info("Adding songs to queue...")
		for _, song := range songs {
			if len(queuedSongs) >= minSongs {
				break
			}
			key := songKey(song.Song)
			if _, ok := existingKeys[key]; ok {
				continue
			}
			out.Write(out.Gray(fmt.Sprintf("  Searching: %s - %s... ", song.Artist, song.Title)))
			shouldPlayNow := !playbackStarted
			result := sonos.AddSongToQueue(ctx, client, room, song.Song, shouldPlayNow)
			if result.Success {
				out.Print(out.Green("found"))
				queuedSongs = append(queuedSongs, song.Song)
				existingKeys[key] = struct{}{}
				if result.TrackID != 0 {
					tried := triedTrackIDsBySong[key]
					if tried == nil {
						tried = map[int]struct{}{}
						triedTrackIDsBySong[key] = tried
					}
					tried[result.TrackID] = struct{}{}
				}
				if shouldPlayNow {
					playbackStarted = true
					out.Success("  -> Playback started!")
					ensureMonitor()
				}
			} else {
				out.Print(out.Red("not found"))
				failedSongs = append(failedSongs, song.Song)
			}
		}
		retryCount++
	}

	out.Print("")
	out.Print(out.Bold(fmt.Sprintf("Queued %d songs on %s", len(queuedSongs), room)))
	if len(failedSongs) > 0 {
		out.Warn(fmt.Sprintf("%d songs not found on Apple Music", len(failedSongs)))
	}
	if !playbackStarted {
		out.Warn("No songs were queued, nothing to play")
	}

	if !monitor {
		out.Info(out.Gray("Queueing complete. Exiting now (use --monitor to keep running)."))
		return Result{
			Room: room, DryRun: false, QueuedSongs: len(queuedSongs), FailedSongs: len(failedSongs),
			PlaybackStarted: playbackStarted, Monitored: false,
		}, nil
	}

	if monitorHandle != nil {
		<-monitorHandle.Done
	}
	return Result{
		Room: room, DryRun: false, QueuedSongs: len(queuedSongs), FailedSongs: len(failedSongs),
		PlaybackStarted: playbackStarted, Monitored: true,
	}, nil
}

File: src/playlist/generator.ts
import chalk from 'chalk';
import {
  generatePlaylistAllProviders,
  generateMoreSongs,
  Song,
  RankedSong,
  APIKeys,
} from '../ai/index.js';
import { generatePlaylistGemini } from '../ai/gemini.js';
import { SonosClient } from '../sonos/client.js';
import { pause } from '../sonos/controls.js';
import { startPlaybackMonitor } from '../sonos/monitor.js';
import {
  clearQueue,
  addSongToQueue,
  addAlternateSongToQueue,
} from '../sonos/queue.js';
import { setTrackReplacement } from '../storage/index.js';

export interface GeneratorOptions {
  keys: APIKeys;
  prompt: string;
  room: string;
  dryRun: boolean;
  client: SonosClient;
  monitor: boolean;
}

const MIN_SONGS = 30;
const SONGS_PER_PROVIDER = 20;
const MAX_RETRIES = 5;
const ALT_RETRY_COOLDOWN_MS = 30_000;
const FAST_START_TIMEOUT_MS = 10_000;

function songKey(song: Song): string {
  return `${song.artist.toLowerCase()}:::${song.title.toLowerCase()}`;
}

async function generateFastStartSongs(
  keys: APIKeys,
  prompt: string
): Promise<Song[]> {
  if (!keys.google) return [];

  const timeout = new Promise<never>((_, reject) => {
    setTimeout(() => {
      reject(new Error(`timed out after ${FAST_START_TIMEOUT_MS / 1000}s`));
    }, FAST_START_TIMEOUT_MS);
  });

  return Promise.race([
    generatePlaylistGemini(keys.google, prompt, 3),
    timeout,
  ]);
}

export async function generateAndPlay(
  options: GeneratorOptions
): Promise<void> {
  const { keys, prompt, room, dryRun, client, monitor } = options;

  let playbackStarted = false;
  let fastStartSong: Song | null = null;
  const existingKeys = new Set<string>();
  const triedTrackIdsBySong = new Map<string, Set<number>>();
  const lastAltAttemptAt = new Map<string, number>();
  let monitorHandle: { stop: () => void; done: Promise<void> } | null = null;

  const handleUnplayable = async (song: Song, trackId: string | null): Promise<void> => {
    const key = songKey(song);
    console.log(
      chalk.yellow(`Detected unplayable track: ${song.artist} - ${song.title}${trackId ? ` (ID: ${trackId}, now blocked)` : ''}`)
    );
    const now = Date.now();
    const lastAttempt = lastAltAttemptAt.get(key) ?? 0;
    if (now - lastAttempt < ALT_RETRY_COOLDOWN_MS) {
      return;
    }
    lastAltAttemptAt.set(key, now);
    const tried = triedTrackIdsBySong.get(key) ?? new Set<number>();
    if (trackId) {
      tried.add(parseInt(trackId, 10));
    }
    const alt = await addAlternateSongToQueue(client, room, song, tried);
    if (alt.success && alt.trackId !== undefined) {
      tried.add(alt.trackId);
      triedTrackIdsBySong.set(key, tried);
      if (trackId) {
        setTrackReplacement(trackId, String(alt.trackId));
      }
      console.log(
        chalk.yellow(
          `Unavailable track replaced with alternate: ${song.artist} - ${song.title}`
        )
      );
    } else {
      console.log(
        chalk.red(
          `No alternate found for: ${song.artist} - ${song.title}`
        )
      );
    }
  };

  const ensureMonitor = (): void => {
    if (!monitor) {
      return;
    }
    if (monitorHandle) {
      return;
    }
    monitorHandle = startPlaybackMonitor(client, room, {
      onUnplayable: handleUnplayable,
      useTimer: false,
    });
    console.log(
      chalk.gray(
        'Monitoring playback for unavailable tracks (Ctrl+C to stop)...'
      )
    );
  };

  let playlistPromise: Promise<Array<Song & { votes?: number }>> | null = null;

  if (!dryRun) {
    console.log(chalk.blue('Generating fast-start songs (Gemini 3 Flash)...'));
    // Start all async operations in parallel for faster startup
    const fastStartPromise = generateFastStartSongs(keys, prompt);
    playlistPromise = generatePlaylistAllProviders(
      keys,
      prompt,
      SONGS_PER_PROVIDER
    );
    const clearPromise = (async () => {
      try {
        await pause(client, room);
      } catch {
        // Ignore pause failures
      }
      await clearQueue(client, room);
    })();

    await clearPromise;

    try {
      const fastStartCandidates = await fastStartPromise;
      if (fastStartCandidates.length === 0) {
        console.log(chalk.yellow('Fast-start unavailable; continuing'));
      } else {
        for (const candidate of fastStartCandidates) {
          // Use playNow=true to switch transport to queue and start playing
          const result = await addSongToQueue(client, room, candidate, true);
          if (!result.success) {
            continue;
          }
          const key = songKey(candidate);
          if (result.trackId !== undefined) {
            const tried = triedTrackIdsBySong.get(key) ?? new Set<number>();
            tried.add(result.trackId);
            triedTrackIdsBySong.set(key, tried);
          }
          existingKeys.add(key);
          playbackStarted = true;
          fastStartSong = candidate;
          console.log(
            chalk.green(
              `  -> Now playing: ${candidate.artist} - ${candidate.title}\n`
            )
          );
          ensureMonitor();
          break;
        }
        if (!playbackStarted) {
          console.log(
            chalk.yellow('Fast-start songs queued but none were playable')
          );
        }
      }
    } catch (error) {
      console.log(
        chalk.yellow(
          `Fast-start skipped (${(error as Error).message}); continuing`
        )
      );
    }
  }

  // Generate playlist from all AI providers
  console.log(
    chalk.blue(`Generating playlist from all providers for: "${prompt}"`)
  );
  let songs: Array<Song & { votes?: number }> = playlistPromise
    ? await playlistPromise
    : await generatePlaylistAllProviders(keys, prompt, SONGS_PER_PROVIDER);
  console.log(
    chalk.green(`Generated ${songs.length} unique songs from all providers\n`)
  );

  // Display initial playlist
  console.log(chalk.bold('Initial playlist:'));
  songs.forEach((song, i) => {
    const votes = song.votes || 1;
    const voteIndicator = votes > 1 ? chalk.yellow(` [${votes}]`) : '';
    console.log(chalk.white(`  ${i + 1}. ${song.artist} - ${song.title}`) + voteIndicator);
  });
  console.log();

  if (dryRun) {
    console.log(chalk.yellow('Dry run - not queueing or playing'));
    return;
  }

  if (playbackStarted) {
    console.log(
      chalk.blue('Fast-start playing; appending generated songs to the queue...')
    );
  }

  // Queue songs and track which ones succeed
  const queuedSongs: Song[] = [];
  const failedSongs: Song[] = [];
  let retryCount = 0;

  while (queuedSongs.length < MIN_SONGS && retryCount < MAX_RETRIES) {
    if (retryCount > 0) {
      console.log(chalk.blue(`\nGenerating more songs (attempt ${retryCount + 1})...`));
      const moreSongs = await generateMoreSongs(
        keys,
        prompt,
        [...queuedSongs, ...failedSongs, ...(fastStartSong ? [fastStartSong] : [])],
        SONGS_PER_PROVIDER
      );
      if (moreSongs.length === 0) {
        console.log(chalk.yellow('Could not generate more unique songs'));
        break;
      }
      songs = moreSongs;
      console.log(chalk.green(`Generated ${songs.length} more songs\n`));
    }

    console.log(chalk.blue('Adding songs to queue...\n'));

    for (const song of songs) {
      const key = songKey(song);
      if (existingKeys.has(key)) {
        continue;
      }
      process.stdout.write(
        chalk.gray(`  Searching: ${song.artist} - ${song.title}... `)
      );
      // Use playNow=true for first song to switch transport to queue
      const shouldPlayNow = !playbackStarted;
      const result = await addSongToQueue(client, room, song, shouldPlayNow);

      if (result.success) {
        console.log(chalk.green('found'));
        queuedSongs.push(song);
        existingKeys.add(key);
        if (result.trackId !== undefined) {
          const tried = triedTrackIdsBySong.get(key) ?? new Set<number>();
          tried.add(result.trackId);
          triedTrackIdsBySong.set(key, tried);
        }
        if (shouldPlayNow) {
          playbackStarted = true;
          console.log(chalk.green('  -> Playback started!\n'));
          ensureMonitor();
        }
      } else {
        console.log(chalk.red('not found'));
        failedSongs.push(song);
      }
    }

    retryCount++;
  }

  // Summary
  console.log();
  console.log(chalk.bold(`Queued ${queuedSongs.length} songs on ${room}`));

  if (failedSongs.length > 0) {
    console.log(chalk.yellow(`\n${failedSongs.length} songs not found on Apple Music`));
  }

  if (!playbackStarted) {
    console.log(chalk.red('\nNo songs were queued, nothing to play'));
  }

  if (!monitor) {
    console.log(
      chalk.gray('Queueing complete. Exiting now (use --monitor to keep running).')
    );
    return;
  }

  if (monitorHandle) {
    await monitorHandle.done;
  }
}
